`ClassCastException` when retrieving composite table entity with `@IdClass`

When I retrieve an entity that models a composite table/relationship between two other tables and uses an `@IdClass` to implement the composite primary, the query fails with a `ClassCastException`. This issue is similar to the failure reported in #1979 (which I also commented on), but not resolved by `2.4.1.Final`.

As an example, I use the following (abbreviated) entities:
```java
@Entity
@Table(name = "entity_a")
public class EntityA {

    @Id
    @Column(name = "id")
    private UUID id;

    @Column(name = "name")
    private String name;

    // Constructor, Getter and Setter
}
```
```java
@Entity
@Table(name = "entity_b")
public class EntityB {

    @Id
    @Column(name = "id")
    private UUID id;

    @Column(name = "name")
    private String name;

    // Constructor, Getter and Setter
}
```

The entities are connected via the following composite table which uses an `@IdClass`:
```java
@Entity
@Table(name = "relationship")
@IdClass(RelationshipId.class)
public class Relationship {

    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "a_id", referencedColumnName = "id")
    private EntityA entityA;

    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "b_id", referencedColumnName = "id")
    private EntityB entityB;

    @Column(name = "dataField")
    private String dataField;

    // Constructor, Getter, Setter, Equal, HashCode
}
```

```java
public class RelationshipId implements Serializable {

    private UUID entityA;
    private UUID entityB;

    // Constructor, Getter, Setter, Equal, HashCode
}
```

The `EntityA` and `EntityB` class can be persisted without issues. The `Relationship` class, can be persisted, but not retrieved. When performing a query such as
```java
session.createQuery("SELECT r FROM Relationship r", Relationship.class).getResultList()
```
the following exception is thrown:
```
11:47:47.417 [vert.x-eventloop-thread-16] INFO  o.h.r.v.impl.DefaultVertxInstance - HR000003: Vert.x instance stopped
Exception in thread "main" java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.hibernate.sql.results.graph.entity.internal.EntityDelayedFetchInitializer cannot be cast to class org.hibernate.reactive.sql.results.graph.ReactiveInitializer (org.hibernate.sql.results
.graph.entity.internal.EntityDelayedFetchInitializer and org.hibernate.reactive.sql.results.graph.ReactiveInitializer are in unnamed module of loader 'app')
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1177)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2341)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:216)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:411)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:382)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:175)
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.reactiveResolveKey(ReactiveNonAggregatedIdentifierMappingInitializer.java:64)
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.reactiveResolveKey(ReactiveNonAggregatedIdentifierMappingInitializer.java:25)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.initializeId(ReactiveEntityInitializerImpl.java:749)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:625)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:605)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:56)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.resolveKey(ReactiveStandardRowReader.java:383)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.lambda$coordinateInitializers$28(ReactiveStandardRowReader.java:376)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$7(CompletionStages.java:410)
        at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:484)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:411)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:382)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.coordinateInitializers(ReactiveStandardRowReader.java:376)        
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.reactiveReadRow(ReactiveStandardRowReader.java:126)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$addToResultsSupplier$5(ReactiveListResultsConsumer.java:141)  
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$consume$1(ReactiveListResultsConsumer.java:96)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:601)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
        at io.vertx.core.Promise.complete(Promise.java:66)
        at io.vertx.core.Promise.handle(Promise.java:51)
        at io.vertx.core.Promise.handle(Promise.java:29)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
        at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:298)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:169)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
        at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
        at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
        at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:114)
        at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:328)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:321)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:388)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)    
        at io.vertx.pgclient.impl.codec.PgDecoder.fireCommandResponse(PgDecoder.java:52)
        at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:137)
        at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:248)
        at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:107)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.hibernate.sql.results.graph.entity.internal.EntityDelayedFetchInitializer cannot be cast to clas
s org.hibernate.reactive.sql.results.graph.ReactiveInitializer (org.hibernate.sql.results.graph.entity.internal.EntityDelayedFetchInitializer and org.hibernate.reactive.sql.results.graph.ReactiveInitializer are in unnamed module of loader 'app')
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.lambda$reactiveResolveKey$1(ReactiveNonAggregatedIdentifierMappingInitializer.java:68)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$2(CompletionStages.java:175)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$7(CompletionStages.java:410)
        at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:484)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215)
        ... 79 more
```

When using Hibernate Reactive `2.3.1.Final`, the entities can be retrieved. In `2.4.0.Final` and `2.4.1.Final`, the exception occurs. Based on the previous ticket, I created the following reproducer of the entities shown above that demonstrates the issue:
[hibernate-reactive-idclass-issue.zip](https://github.com/user-attachments/files/17203194/hibernate-reactive-idclass-issue.zip)


You can execute it with `./gradlew run`. If you want to inspect the behaviour in the previous version, you can adjust the `ext` block in the `build.gradle`.

Best regards,
Janek

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingFetch.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingFetch.java
@@ -7,6 +7,8 @@ package org.hibernate.reactive.sql.results.graph.embeddable.internal;
 
 import org.hibernate.engine.FetchTiming;
 import org.hibernate.metamodel.mapping.NonAggregatedIdentifierMapping;
+import org.hibernate.metamodel.mapping.internal.ToOneAttributeMapping;
+import org.hibernate.reactive.metamodel.mapping.internal.ReactiveToOneAttributeMapping;
 import org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityFetchJoinedImpl;
 import org.hibernate.spi.NavigablePath;
 import org.hibernate.sql.results.graph.AssemblerCreationState;
@@ -50,8 +52,15 @@ public class ReactiveNonAggregatedIdentifierMappingFetch extends ReactiveEmbedda
 			boolean selected,
 			String resultVariable,
 			DomainResultCreationState creationState) {
+		// I don't think this is the right approach.
+		// Fetchable should already be an instance of ReactiveToOneAttributeMapping
+		// (if it's an instance of ToOneAttributeMapping). But I don't think there is a way right now
+		// to make it happen without changing ORM.
+		Fetchable reactiveFetchable = fetchable instanceof ToOneAttributeMapping && !( fetchable instanceof ReactiveToOneAttributeMapping )
+				? new ReactiveToOneAttributeMapping( (ToOneAttributeMapping) fetchable )
+				: fetchable;
 		Fetch fetch = super.generateFetchableFetch(
-				fetchable,
+				reactiveFetchable,
 				fetchablePath,
 				fetchTiming,
 				selected,

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneLazyIdClassTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneLazyIdClassTest.java
@@ -1,0 +1,320 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+import org.junit.jupiter.api.Test;
+
+import io.vertx.junit5.Timeout;
+import io.vertx.junit5.VertxTestContext;
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.FetchType;
+import jakarta.persistence.Id;
+import jakarta.persistence.IdClass;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Timeout(value = 10, timeUnit = MINUTES)
+public class ManyToOneLazyIdClassTest extends BaseReactiveTest {
+
+	@Override
+	protected Collection<Class<?>> annotatedEntities() {
+		return List.of( Relationship.class, EntityA.class, EntityB.class );
+	}
+
+	@Test
+	public void test(VertxTestContext context) {
+		final UUID aId = UUID.randomUUID();
+		final EntityA entityA = new EntityA( aId, "testA" );
+		final UUID bId = UUID.randomUUID();
+		final EntityB entityB = new EntityB( bId, "testB" );
+		final Relationship relationship = new Relationship( entityA, entityB, "testRelationship" );
+
+		test( context, getMutinySessionFactory()
+				.withTransaction( s -> s.persistAll( entityA, entityB, relationship ) )
+				.chain( () -> getMutinySessionFactory().withTransaction( s -> s
+						.find( EntityA.class, aId )
+						.invoke( entity -> assertThat( entity ).isEqualTo( entityA ) )
+				) )
+				.chain( () -> getMutinySessionFactory().withTransaction( s -> s
+						.find( EntityB.class, bId )
+						.invoke( entity -> assertThat( entity ).isEqualTo( entityB ) )
+				) )
+				.chain( () -> getMutinySessionFactory().withTransaction( s -> s.find(
+						Relationship.class,
+						new RelationshipId( aId, bId )
+				) ) )
+		);
+	}
+
+	@Entity
+	@Table(name = "entity_a")
+	public static class EntityA {
+
+		@Id
+		@Column(name = "id")
+		private UUID id;
+
+		@Column(name = "name")
+		private String name;
+
+		public EntityA(UUID id, String name) {
+			this.id = id;
+			this.name = name;
+		}
+
+		public EntityA() {
+		}
+
+		public UUID getId() {
+			return id;
+		}
+
+		public void setId(UUID id) {
+			this.id = id;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public void setName(String name) {
+			this.name = name;
+		}
+
+		@Override
+		public String toString() {
+			return id + ":" + name;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( o == null || getClass() != o.getClass() ) {
+				return false;
+			}
+			EntityA entityA = (EntityA) o;
+			return Objects.equals( id, entityA.id ) && Objects.equals( name, entityA.name );
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash( id, name );
+		}
+	}
+
+	@Entity
+	@Table(name = "entity_b")
+	public static class EntityB {
+
+		@Id
+		@Column(name = "id")
+		private UUID id;
+
+		@Column(name = "name")
+		private String name;
+
+		public EntityB(UUID id, String name) {
+			this.id = id;
+			this.name = name;
+		}
+
+		public EntityB() {
+		}
+
+		public UUID getId() {
+			return id;
+		}
+
+		public void setId(UUID id) {
+			this.id = id;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public void setName(String name) {
+			this.name = name;
+		}
+
+		@Override
+		public String toString() {
+			return id + ":" + name;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( o == null || getClass() != o.getClass() ) {
+				return false;
+			}
+			EntityB entityB = (EntityB) o;
+			return Objects.equals( id, entityB.id ) && Objects.equals( name, entityB.name );
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash( id, name );
+		}
+	}
+
+	@Entity( name = "Relationship")
+	@Table(name = "relationship")
+	@IdClass(RelationshipId.class)
+	public static class Relationship {
+
+		@Id
+		@ManyToOne(fetch = FetchType.LAZY)
+		@JoinColumn(name = "a_id", referencedColumnName = "id")
+		private EntityA entityA;
+
+		@Id
+		@ManyToOne(fetch = FetchType.LAZY)
+		@JoinColumn(name = "b_id", referencedColumnName = "id")
+		private EntityB entityB;
+
+		@Column(name = "dataField")
+		private String dataField;
+
+		public Relationship(EntityA entityA, EntityB entityB, String dataField) {
+			this.entityA = entityA;
+			this.entityB = entityB;
+			this.dataField = dataField;
+		}
+
+		public Relationship() {
+		}
+
+		public EntityA getEntityA() {
+			return entityA;
+		}
+
+		public void setEntityA(EntityA entityA) {
+			this.entityA = entityA;
+		}
+
+		public EntityB getEntityB() {
+			return entityB;
+		}
+
+		public void setEntityB(EntityB entityB) {
+			this.entityB = entityB;
+		}
+
+		public String getDataField() {
+			return dataField;
+		}
+
+		public void setDataField(String dataField) {
+			this.dataField = dataField;
+		}
+
+		@Override
+		public String toString() {
+			return entityA + "-" + entityB + ":" + dataField;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if ( this == obj ) {
+				return true;
+			}
+			if ( !( obj instanceof Relationship ) ) {
+				return false;
+			}
+			Relationship other = (Relationship) obj;
+			if ( entityA == null
+					|| other.getEntityA() == null
+					|| entityB == null
+					|| other.getEntityB() == null
+			) {
+				return false;
+			}
+			return entityA.equals( other.getEntityA() )
+					&& entityB.equals( other.getEntityB() );
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash( entityA, entityB );
+		}
+	}
+
+	public static class RelationshipId {
+
+		private UUID entityA;
+		private UUID entityB;
+
+		public RelationshipId() {
+		}
+
+		public RelationshipId(UUID entityA, UUID entityB) {
+			this.entityA = entityA;
+			this.entityB = entityB;
+		}
+
+		public UUID getEntityA() {
+			return entityA;
+		}
+
+		public void setEntityA(UUID entityA) {
+			this.entityA = entityA;
+		}
+
+		public UUID getEntityB() {
+			return entityB;
+		}
+
+		public void setEntityB(UUID entityB) {
+			this.entityB = entityB;
+		}
+
+		@Override
+		public String toString() {
+			return entityA + ":" + entityB;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if ( this == obj ) {
+				return true;
+			}
+			if ( !( obj instanceof RelationshipId ) ) {
+				return false;
+			}
+			RelationshipId other = (RelationshipId) obj;
+			if ( entityA == null
+					|| other.getEntityA() == null
+					|| entityB == null
+					|| other.getEntityB() == null
+			) {
+				return false;
+			}
+			return entityA.equals( other.getEntityA() )
+					&& entityB.equals( other.getEntityB() );
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash( entityA, entityB );
+		}
+	}
+
+}

